hydro_deploy/hydro_deploy_integration/src/multi_connection.rs

```diff
@@ -0,0 +1,351 @@
+use std::collections::HashMap;
+use std::io;
+use std::marker::PhantomData;
+use std::ops::DerefMut;
+use std::pin::Pin;
+use std::task::{Context, Poll, ready};
+
+use futures::{Sink, SinkExt, Stream, StreamExt};
+use tokio::net::TcpListener;
+#[cfg(unix)]
+use tokio::net::UnixListener;
+use tokio::sync::mpsc;
+use tokio_util::codec::{Decoder, Encoder, Framed};
+
+use crate::{AcceptedServer, BoundServer, Connected, ConnectedSourceSink, Connection};
+
+#[expect(clippy::type_complexity, reason = "Dynamic generic stream handling")]
+pub struct ConnectedMultiConnection<I, O, C: Decoder<Item = I> + Encoder<O>> {
+    source: Pin<Box<dyn Stream<Item = Result<(u64, I), <C as Decoder>::Error>> + Send + Sync>>,
+    sink: Pin<Box<dyn Sink<(u64, O), Error = <C as Encoder<O>>::Error> + Send + Sync>>,
+}
+
+impl<
+    I: 'static,
+    O: Send + Sync + 'static,
+    C: Decoder<Item = I> + Encoder<O> + Send + Sync + Default + 'static,
+> Connected for ConnectedMultiConnection<I, O, C>
+{
+    fn from_defn(pipe: Connection) -> Self {
+        match pipe {
+            Connection::AsServer(AcceptedServer::MultiConnection(bound_server)) => {
+                let (source, sink) = MultiConnectionStream::create::<I, O, C>(*bound_server);
+                ConnectedMultiConnection {
+                    source: Box::pin(source),
+                    sink: Box::pin(sink),
+                }
+            }
+            _ => panic!("Cannot connect to a non-multi-connection pipe as a multi-connection"),
+        }
+    }
+}
+
+impl<
+    I: Send,
+    O: Send + Sync + 'static,
+    C: Decoder<Item = I> + Encoder<O> + Send + Sync + Default + 'static,
+> ConnectedSourceSink for ConnectedMultiConnection<I, O, C>
+{
+    type Output = (u64, I);
+    type OutError = <C as Decoder>::Error;
+    type Stream =
+        Pin<Box<dyn Stream<Item = Result<Self::Output, <C as Decoder>::Error>> + Send + Sync>>;
```

Comment:
Thought experiment: can this be not `Box<dyn ...>`

Reply:
Fixed!